### ISSUE TITLE
i18next & common questions

### ISSUE THREAD

**Issue by 7iomka** [UnlyEd/next-right-now#46]
Hello! Thank you very much for your work!
It is possible to continue usage of i18next but without locize vendor?
Advise steps for implementation please!

PS: I tried using next-i18next but their configuration uses `getInitialProps` in `_app` and this **disables static optimization**. The developers themselves [refuse](https://github.com/isaachinman/next-i18next/issues/586#issuecomment-624810221) to change anything in this direction. 

Then I found your repository and studied the approaches you proposed!
Thanks for any help and awesome ideas in context of problem solving with latest next.js!

**Comment by Vadorequest** [comment]
Hello @7iomka :)

Yes, it's totally possible to do without Locize, even though no official preset was released without Locize built-in so far. (I'll do that someday, but want to polish current SSG support work in progress before)

Also, I invite you to check the current PR regarding **SSG** support:
https://github.com/UnlyEd/next-right-now/pull/42

This branch provides full SSG support using the latest Next.js version.
I advise you start from it, and remove locize.

To remove locize, the easiest approach would be to override the following:
- https://github.com/UnlyEd/next-right-now/blob/d1204230db011189ff262ee3e9284a024f9a47f0/src/utils/nextjs/SSG.ts#L43 Currently loads translations from Locize API, instead you should load your i18next translation your own way (filesystem?) 
- https://github.com/UnlyEd/next-right-now/blob/d1204230db011189ff262ee3e9284a024f9a47f0/src/components/appBootstrap/MultiversalAppBootstrap.tsx#L60 takes i18nTranslations (currently fetched from Locize API) and init the i18next with it, there is nothing to change here, it's just for clarity :)
- https://github.com/UnlyEd/next-right-now/blob/a7307e0638aa4c44cab70c5188a64df17f09277f/src/utils/i18n/i18nextLocize.ts#L371 Remove all Locize plugins (line 342 and below)

I believe those small manipulations should be enough to get rid of Locize, using the filesystem may be the easiest way. Note that you need to respect the i18next spec regarding data format though.

Here is an example of `i18nTranslations`:

```json
{
  "en": {
    "common": {
      "examples": {
        "i18n": {
          "dynamicPluralTranslation": "We found {{count}} solution for you.",
          "dynamicPluralTranslation_plural": "We found {{count}} solutions for you.",
          "dynamicTranslation": "Dynamic content: <1>{{uuid}}</1>",
          "pluralTranslation": "Translation with pluralisation",
          "pluralTranslation_plural": "Translation with pluralisations",
          "simpleTranslation": "Simple translation"
        }
      },
      "footer": {
        "terms": {
          "link": "Terms",
          "text": "All rights reserved"
        }
      },
      "nav": {
        "adminSite": {
          "link": "Admin site"
        },
        "examplesPage": {
          "link": "Examples"
        },
        "githubDocPage": {
          "link": "Documentation"
        },
        "githubPage": {
          "link": "Source code"
        },
        "indexPage": {
          "link": "Home"
        },
        "productsPage": {
          "link": "Products"
        },
        "termsPage": {
          "link": "Terms"
        }
      }
    }
  }
}
```

**Comment by Vadorequest** [comment]
@jajourda FYI this could be useful to you too :)

**Comment by 7iomka** [comment]
@Vadorequest Thanks you very much! I will check proposed steps a little later.
I'm worked with my own branch, because I don’t need much of what your branch offers (analytics, logging, multiple utils)
But in order to borrow your approaches to reorganizing code in `_app`, I now need to study a lot of your code, leaving only what I need. Ultimately, when this routine work is over, I hope that I get a working code :)
Next.js did a great job of separating rendering options, but with regard to the implementation of user code that users had previously initialized in `_app` via `getInitialProps`, it is unclear how to "quickly" abandon this.

Hopefully in the future next will make it possible to determine the type of page rendering in `_app`.
And I like your suggestions on this: 
<details>
  <summary>Spoiler</summary>

  ```
 { mode: 'SSG', origin: 'server', servedBy: 'server' } - Page served dynamically (getStaticProps, server build, when working locally)
{ mode: 'SSG', origin: 'server', servedBy: 'browser' } - Page served dynamically (getStaticProps, client re-render, when working locally)
{ mode: 'SSG', origin: 'browser', servedBy: 'browser'} - Page served statically (getStaticProps, static bundle, client-side)
{ mode: 'SSR', origin: 'server', servedBy: 'server'} - Page served dynamically (getInitialProps, server initial render)
{ mode: 'SSR', origin: 'browser', servedBy: 'browser'} - Page served dynamically (getInitialProps, client re-render/render (CSR))
{ mode: 'SSR' (SSP ?), origin: 'server', servedBy: 'server'} - Page served dynamically (getServerSideProps, called from server directly (e.g: page refresh)
{ mode: 'SSR' (SSP ?), origin: 'browser', servedBy: 'server'} - Page served dynamically (getServerSideProps, called from browser and proxied through API (e.g: client-side navigation to a page using getServerSideProps)
  ```
</details>

**Comment by Vadorequest** [comment]
Yes, it does come with a lot of built-in stuff, at some point I may release lighter presets. Current preset is actually based on our own internal requirements, hence the complexity.

I should make a RFC about this suggestion actually... Timer (from Next.js core team) asked me ahah

**Comment by 7iomka** [comment]
> Yes, it does come with a lot of built-in stuff, at some point I may release lighter presets. Current preset is actually based on our own internal requirements, hence the complexity.
> 
> I should make a RFC about this suggestion actually... Timer (from Next.js core team) asked me ahah

Yes, you did a lot of work to create such an assembly with such a set of different tools for different tasks. For newcomers (such as me) it is very difficult to tear out some functional isolated. For example, I am not familiar with architecture of now deployment, and I'm a little confused about how I define this env variable for various node_env (I have this question because I see you use multiple checking rules like `if (process.env.APP_STAGE === 'development')`.
I started reading the documentation for next.js
https://nextjs.org/docs/basic-features/environment-variables#exposing-environment-variables
Do I understand correctly what I need to create (?)
1) **.env.development**
with 
`APP_STAGE=development`
2) **.env.production**
with 
`APP_STAGE=production`
3) **.env**
with my common needed and non-sensitive env variables
4) .env.local / .env.development.local / .env.production.local (git ignored)
with sensitive env variables (like api keys, etc.)

5) in **next.config.js** in `env` field
only env variables computed by other env variables or calculated by other source of data
<details>
  <summary>Like this</summary>

  ```
 // Non duplicated environment variables (automatically resolved, must not be specified in the .env.build file)
    BUILD_TIME: date.toString(),
    BUILD_TIMESTAMP: +date,
    APP_NAME: packageJson.name,
    APP_VERSION: packageJson.version,
    UNLY_SIMPLE_LOGGER_ENV: process.env.APP_STAGE, // Used by @unly/utils-simple-logger - Fix missing staging logs because it believes we're in production
  ```
</details>


And after this next.js will handle and will use needed env variables for each stage (development or production). One different with your configs is that with this approach we won't have `staging` stage.

**Comment by Vadorequest** [comment]
Actually, you only need to create what's recommended in https://unlyed.github.io/next-right-now/getting-started/quick-start, which is `.env.build`.

If you deploy on Vercel, only the now.*.json files are used in staging/production. The .env.build file is only used locally.

Reading https://unlyed.github.io/next-right-now/concepts/env-and-stages should help, too.

**Comment by 7iomka** [comment]
> Actually, you only need to create what's recommended in https://unlyed.github.io/next-right-now/getting-started/quick-start, which is `.env.build`.
> 
> If you deploy on Vercel, only the now.*.json files are used in staging/production. The .env.build file is only used locally.
> 
> Reading https://unlyed.github.io/next-right-now/concepts/env-and-stages should help, too.

Ok! Thanks.
One last question (may be need to create another issue with questions like this).

it's possible to rewrite _app as functional component (currently you use class component)?
I try this
```
const MultiversalPageEntryPoint = ({ Component, pageProps }: MultiversalAppBootstrapProps) => {
   return <MultiversalAppBootstrap {...**WHAT PROPS to pass there???**} />
}
```

**Comment by Vadorequest** [comment]
No, it's not possible to use a class component (AFAIK), because it has to extends the native `import NextApp from 'next/app';`, and relies on it for error handling and more.

**Comment by 7iomka** [comment]
> No, it's not possible to use a class component (AFAIK), because it has to extends the native `import NextApp from 'next/app';`, and relies on it for error handling and more.

My question is not closely related to your assembly.
I want to implement my `_app` as a functional component (in accordance with the recommended use case in the documentation).
In my assembly I don't use error handling and I don't need `componentDidCatch`
So I don't know what type of props I need to pass to MultiversalAppBootstrap component

**Comment by Vadorequest** [comment]
I suggest you ask on the Next.js community for this, but my recommendation is to keep it as it. You should look at what `import NextApp from 'next/app';` does underneath, I'm fairly sure Next.js app **can't work without out it**. It's necessary, IMHO.

**Comment by 7iomka** [comment]
@Vadorequest , Official docs example https://nextjs.org/docs/basic-features/typescript#custom-app - this is functional component and with this all works fine
My question is how to integrate your component MultiversalAppBootstrap in _app if it is a functional component?
![image](https://user-images.githubusercontent.com/18473137/81695241-0a185600-946b-11ea-9d8b-37f5a15c1c51.png)

**Comment by Vadorequest** [comment]
Oh, interesting, they've changed that. Thanks, I'll use a functional component too then :)

Based on their examples:

```tsx
function MyApp(props: AppProps) {
  return <MultiversalAppBootstrap {...props} />
}
```

This should do it.

**Comment by 7iomka** [comment]
> Oh, interesting, they've changed that. Thanks, I'll use a functional component too then :)
> 
> Based on their examples:
> 
> ```tsx
> function MyApp(props: AppProps) {
>   return <MultiversalAppBootstrap {...props} />
> }
> ```
> 
> This should do it.

There is still work to be done on indicating the correct type, because now it shows an error in my assembly.
Perhaps these are some of my improvements. Do you have any errors with this use?
My current workaround is ` return <MultiversalAppBootstrap {...(props as any)} />;`

**Comment by 7iomka** [comment]
Oh
I found workaround: just import `Props` type from `MultiversalAppBootstrap` component
and pass it to `MyApp(props: Props)` - it fixes error with typing but I don't know if this is correct way)

**Comment by 7iomka** [comment]
Another small question about i18n implementation: do you have any thoughts on the implementation of your TODO ?
https://github.com/UnlyEd/next-right-now/blob/v1-hyb-mst-aptd-gcms-lcz-sty/src/utils/i18n/i18n.ts#L32

**Comment by Vadorequest** [comment]
Nope, but I'll address that before considering this preset "done". (please open different issues 😉)

**Comment by 7iomka** [comment]
> please open different issues

I just renamed this title for consistence.

Another hot question about how works i18next integration in this assembly:
how you make changes with url programmatically? I mean something like `Router.push('/products')`
Currently I see you don't have an customized Router instance to do that without page reload (in `utils/app/router` I just found `resolveI18nRoute` used in customized component `Link`.
Thanks.

**Comment by 7iomka** [comment]
Another question- it is possible to make to switch locale to another one without full page reload?
If for us `i18nextInstance` available inside our pages _(yes by the way with this you forgot to add to types `MultiversalAppBootstrapPageProps`)_
you can make language-switcher and use code like this: 
`<button type="button" onClick={() => i18nextInstance.changeLanguage('fr')}>`
This will change locale to 'fr' but with current implementation we have 2 problems
1) if we will use i18next without locize (as me), we don't have access to `fr` (i18nTranslations includes only previous locale in key).
2) old locale still present in browser url, and we need to replace it.

I tried to replace `i18nextLocizeBackend` with [i18nextHttpBackend](https://github.com/i18next/i18next-http-backend) but it do nothing :(

**Comment by Vadorequest** [comment]
@7iomka You must use the native Next.js router for programmatic usage.
`const router: NextRouter = useRouter();`

---

It's not possible to change the language this way, you need to use the router to change the file being used. But I'll actually implement that on the PR so you can also just wait for it!

**Comment by Vadorequest** [comment]
I've changed the implementation of language switching in the footer. It doesn't do a full page refresh anymore.

https://nrn-v1-hyb-mst-aptd-gcms-lcz-sty-c1-i5bbeso8b.now.sh/fr/examples

**Comment by 7iomka** [comment]
@Vadorequest, see please my 1) question
https://github.com/UnlyEd/next-right-now/issues/47#issuecomment-629370380



> I've changed the implementation of language switching in the footer. It doesn't do a full page refresh anymore.
> 
> https://nrn-v1-hyb-mst-aptd-gcms-lcz-sty-c1-i5bbeso8b.now.sh/fr/examples

I tested this example, it works nice except one strange behavior:
1) Open page with `en` locale: https://nrn-v1-hyb-mst-aptd-gcms-lcz-sty-c1-i5bbeso8b.now.sh/en/examples
2) disable javascript on the page (chrome dev-tools => ctrl/cmd + shift + p => type `disable javascript`)
3) reload page
Instead of `en` locale you will see page in french! Why?) It's my mistake or something wrong in code logic?

**Comment by Vadorequest** [comment]
I don't reproduce this, if I disabled JS then nothing happens when I click on the flag image.

Here is a video of the behaviour, seems to behave as expected.
https://youtu.be/F9kWFgKMrT8

**Comment by 7iomka** [comment]
> I don't reproduce this, if I disabled JS then nothing happens when I click on the flag image.
> 
> Here is a video of the behaviour, seems to behave as expected.
> https://youtu.be/F9kWFgKMrT8

you were inattentive)


> 3\. **reload** page

don't try to toggle language without js) 

try steps from above and you will se page with `en` but in french after manually reload (got fro server)
![image](https://user-images.githubusercontent.com/18473137/82081640-57a4f500-96ef-11ea-9036-bee4fbceaae5.png)

**Comment by 7iomka** [comment]
Another improvement idea - include query parameters to not be removed after locale changed.
```ts
export const i18nRedirect = (locale, router: NextRouter, pageReload = false): void => {
  const newUrl = `${router.pathname.replace('[locale]', locale)}`;
  const { locale: queryLocale, ...routerQueryParamsExceptLocale } = router.query || {};

  if (pageReload) {
    window.location.href = newUrl;
  } else {
    router.push(
      {
        pathname: router.pathname,
        query: router.query,
      },
      {
        pathname: newUrl,
        query: {
          ...routerQueryParamsExceptLocale,
        },
      },
    );
  }
};
```

**Comment by Vadorequest** [comment]
Okay, I see. 

FYI you can use https://chrome.google.com/webstore/detail/screencastify-screen-vide/mmeijimgabbpbgpdklnllpncmdofkcpn for bug reporting, it's great. :)

**Comment by Vadorequest** [comment]
It's due to a caching issue caused by the i18nextLocize singleton, I'll fix it quickly.

**Comment by Vadorequest** [comment]
Fixed by https://github.com/UnlyEd/next-right-now/pull/42/commits/386ac66e4afe7f5f433048764e630a40438d0fe7

demo https://nrn-v1-hyb-mst-aptd-gcms-lcz-sty-c1-d66yedqcz.now.sh/fr

**Comment by 7iomka** [comment]
> Fixed by [386ac66](https://github.com/UnlyEd/next-right-now/commit/386ac66e4afe7f5f433048764e630a40438d0fe7)
> 
> demo https://nrn-v1-hyb-mst-aptd-gcms-lcz-sty-c1-d66yedqcz.now.sh/fr

Nice! 👍 
One question about [386ac66](https://github.com/UnlyEd/next-right-now/commit/386ac66e4afe7f5f433048764e630a40438d0fe7)
you overwrite globalInstance of i18n instance instead to `.addResourceBundle`, - it is normal practice?
In my use-case I will have live-edit mode for some page, and for the end-user, switching the language to change translated entities will be a frequent and trivial task.
I18next have possibility to store multiple loaded locales and override if needed
For example: https://www.i18next.com/how-to/add-or-load-translations#add-after-init
What are the benefits of overwriting an entire instance?
I suppose it was necessary when the page was completely reloaded, but now you have changed this logic.

**Comment by Vadorequest** [comment]
I didn't look into the i18next API and wasn't aware of `addResourceBundle`.

There is no benefit about overriding the whole instance, but I guess this way I'm sure I don't run into caching issues.

Also, it's not really an issue with the way it currently works, I don't see any real benefit about not overriding it. (maybe would be faster, but it's so fast already it wouldn't be humanly perceptible)
